test/UpdaterAndFilterEngineCreation.cpp

```diff
@@ -0,0 +1,172 @@
+
+namespace
+{
+  class UpdaterAndFilterEngineCreationTest : public BaseJsTest
+  {
+  protected:
+
+    static const size_t COUNT = 100;
+    const std::string PROP_NAME = "patternsbackupinterval";
+    std::vector<std::thread> threadsVector;
+    Updater* updaterAddrArray[COUNT];
+    FilterEngine* filterAddrArray[COUNT];
+    DelayedWebRequest::SharedTasks webRequestTasks;
+    DelayedTimer::SharedTasks timerTasks;
+
+    void SetUp()
+    {
+      LazyFileSystem* fileSystem;
+      ThrowingPlatformCreationParameters platformParams;
+      platformParams.logSystem.reset(new LazyLogSystem());
+      platformParams.timer = DelayedTimer::New(timerTasks);
+      platformParams.fileSystem.reset(fileSystem = new LazyFileSystem());
+      platformParams.webRequest = DelayedWebRequest::New(webRequestTasks);
+      platform.reset(new Platform(std::move(platformParams)));
+      threadsVector.reserve(COUNT);
+      std::uninitialized_fill(updaterAddrArray, updaterAddrArray + COUNT, nullptr);
+      std::uninitialized_fill(filterAddrArray, filterAddrArray + COUNT, nullptr);
+    }
+
+    void CallGetUpdaterSimultaneously()
+    {
+      CallGetSimultaneously(true, false);
+    }
+
+    void CallGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(false, true);
+    }
+
+    void CallGetUpdaterAndGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(true, true);
+    }
+
+  private:
+
+    void CallGetSimultaneously(bool isUpdater, bool isFilterEngine)
+    {
+      for (size_t idx = 0; idx < COUNT; ++idx)
+        threadsVector.push_back(
+          std::thread([this, idx, isUpdater, isFilterEngine]() -> void {
+            Updater* updaterAddr = nullptr;
+            FilterEngine* filterEngineAddr = nullptr;
+            if (isUpdater && isFilterEngine)
+            {
+              // some randomization in order of calling gets
+              if (idx % 2)
+              {
+                updaterAddr = &(platform->GetUpdater());
+                filterEngineAddr = &(platform->GetFilterEngine());
+              }
+              else
+              {
+                filterEngineAddr = &(platform->GetFilterEngine());
+                updaterAddr = &(platform->GetUpdater());
+              }
+            }
+            else if (isUpdater)
+              updaterAddr = &(platform->GetUpdater());
+            else if (isFilterEngine)
+              filterEngineAddr = &(platform->GetFilterEngine());
+
+            if (updaterAddr != nullptr)
+              updaterAddrArray[idx] = updaterAddr;
+
+            if (filterEngineAddr != nullptr)
+              filterAddrArray[idx] = filterEngineAddr;
+          }));
+
+      // join the threads
+      for (auto& th: threadsVector)
+        if (th.joinable())
+          th.join();
+    }
+  };
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestFilterEngineSingleInstance)
+{
+    CallGetFilterEngineSimultaneously();
+    FilterEngine* filterEngineAddr = filterAddrArray[0];
```

Comment:
Strictly speaking it would be good to test just in case that the value of the first element is not `nullptr` since we expect it anyway.

Reply:
Good point.